docs/docs/validating-plugin-options.md

```diff
@@ -0,0 +1,230 @@
+         optionA: false,
+         message: "string",
+         optionB: true,
+       }
+       const { isValid, errors } = testPluginOptionsSchema(
+         pluginOptionsSchema,
+         options
+       )
+
+       expect(isValid).toBe(true)
+       expect(errors).toEqual([])
+     })
+   })
+   ```
+
+## Best practices for option schemas
+
+The [Joi API documentation](https://joi.dev/api/) is a great reference to use while working on a `pluginOptionsSchema` as it shows all the available types and methods.
+
+Here are some specific Joi best practices for `pluginOptionsSchema`s:
+
+- [Add descriptions](#add-descriptions)
+- [Set default options](#set-default-options)
+- [Validate external access](#validate-external-access) where necessary
+- [Add custom error messages](#add-custom-error-messages) where useful
+- [Deprecate options](#deprecating-options) in a major version release rather than removing them
+
+### Add descriptions
+
+Make sure that every option and field has a `.description()` explaining its purpose. While this is not immediately visible, there might be tooling in the future that auto-generates plugin option documentation from the schema.
+
+### Set default options
+
+You can use the `.default()` method to default a value for an option. For example with `gatsby-plugin-console-log`, this is how it can default the `message` option to `"default message"` in case users do not pass their own `message`:
+
+```javascript:title=plugins/gatsby-plugin-console-log/gatsby-node.js
+exports.pluginOptionsSchema = ({ Joi }) => {
+  return Joi.object({
+    optionA: Joi.boolean().required().description(`Enables optionA.`),
+    message: Joi.string()
+      .default(`default message`) // highlight-line
+      .description(`The message logged to the console.`),
+    optionB: Joi.boolean().description(`Enables optionB.`),
+  })
+}
+```
+
+Accessing `pluginOptions.message` would then log `"default message"` in all plugin APIs if the user does not supply their own value:
+
+```javascript:title=plugins/gatsby-plugin-console-log/gatsby-node.js
+exports.onPreInit = (_, pluginOptions) => {
+  console.log(
+    `logging: "${pluginOptions.message}" to the console` // highlight-line
+  )
+}
+```
+
+### Validating external access
+
+Some plugins (particularly source plugins) query external APIs. With the `.external()` method, they can asynchronously validate that users have access to the API to provide a better experience if they pass invalid secrets.
+
+For example, this is how the [Contentful source plugin](/plugins/gatsby-source-contentful/) ensures the user has access to the space they are trying to query:
+
+```javascript:title=gatsby-source-contentful/gatsby-node.js
+exports.pluginOptionsSchema = ({ Joi }) => {
+  return Joi.object({
+    accessToken: Joi.string().required(),
+    spaceId: Joi.string().required(),
+    // ...more options here...
+  }).external(async pluginOptions => {
+    try {
+      await contentful
+        .createClient({
```

Comment:
Keen eyes will spot that this is not the actual code used in the Contentful plugin. That is on purpose, as the actual code uses a vanilla `fetch` call with headers etc. and is more verbose, i.e. makes for a bad example as it hides the true intent of the function.

I figured the contentful SDK API would make the example clearer!

Reply:
It might be helpful to put that caveat in here, so that people don't use this as a source of truth re: how Contentful works. It might be as simple as just changing the paragraph introducing the example to something like, "For example, this is how the Contentful source plugin schema might look if they wanted to validate using [whatever thing they would use instead of `fetch`]"